tools/rum-models-generator/Sources/RUMModelsGeneratorCore/RUM/RUMSwiftTypeTransformer.swift

```diff
@@ -198,6 +217,18 @@ internal class RUMSwiftTypeTransformer {
             fixedName = "RUMOperatingSystem"
         }
 
+        if fixedName == "Display" {
+            fixedName = "RUMDisplay"
+        }
+
+        // Since https://github.com/DataDog/rum-events-format/pull/57 `action.id` can be either
+        // single `String` or an array of `[String]`. This is handled by generating Swift enum with
+        // two cases and different associated types. To not duplicate generated code in each nested
+        // context we generate single root type: `RUMActionID`.
+        if fixedName == "ID", let parentStructName = (context.parent as? SwiftStruct)?.name, parentStructName == "action" {
+            fixedName = "RUMActionID"
+        }
```

Comment:
/question
It would make sense that all `$ref` are not nested, while other types are, no? instead of hard coding this mapping in the generator.

Reply:
I see what you mean. It more relates to this definition:
```swift
/// Types which will shared between all input `types`. Sharing means detaching those types from nested declaration
/// and putting them at the root level of the resultant `types` array, so the type can be printed without being nested.
private let sharedTypeNames = [
   "RUMConnectivity",
   "RUMUser",
   "RUMMethod",
   "RUMEventAttributes",
   "RUMCITest",
   "RUMDevice",
   "RUMOperatingSystem",
   "RUMDisplay",
   "RUMActionID",
]
```

Now, when comparing this to [`_common-schema.json`](https://github.com/DataDog/rum-events-format/blob/master/schemas/rum/_common-schema.json) (a common `$ref` in multiple event schemas), indeed most of these objects are part of it and could be inferred automatically. But some other are not (e.g. `RUMEventAttributes` or `RUMMethod`), similar to few primitive values (e.g. `service` or `version`), so it would still require manual patching with opt-ins or opt-outs declarations.

I will look closer at this when working on Session Replay models. Compared to RUM events, these use way more `$refs` and automated inferring could work better.

Last, the `RUMSwiftTypeTransformer` is only an optional add-on to the entire JSON schema <> Swift generation system implemented by `rum-models-generator`. It only decorates intermediate Swift schema before it is passed to `SwiftPrinter`, but everything would work perfectly well without it (just the names would be weird and structure not so convenient). Because it is well decoupled, implementing few decorative hacks in there, like `sharedTypeNames` or `fixedName` is not a big problem IMO.